### DISCUSSION TITLE
[Feature Request] PGP sign and/or encrypt E-Mails

### DISCUSSION THREAD

**Discussion by DunklerPhoenix** [dani-garcia/vaultwarden#2418] [Ideas]
Heho
It would be great if E-Mails can be signed or encrypted by an PGP key. The user should decide which mail-type should be normal, signed or encrypted.

Maybe this is not possible, because Bitwarden itself doesn't support the options.

**Comment by BlackDex** [comment]
That would be very difficult to configure i think.
There isn't an option for this within the user interface to add or modify there public PGP key which is needed for the encryption.
Without this the server would need to try and guess where they need to download the key from if that is possible at all.

Facebook for example has this feature. But there you need to upload your public key so that they can use that with there key to encrypt the message.

So, i do not think this is a feasible option to try and implement.

**Comment by DunklerPhoenix** [comment]
I thought some hours about a solution, but everything is to complicated for an average user. The public key wouldn't be the greatest difficult, but more the options for changing the settings per mail type.

**Reply by BlackDex** [reply]
Well, since we do not want to modify the web-vault it is difficult.
You could try to create a totally separated environment like the `/admin` interface. But that would also be difficult to manage

**Reply by tamaskan** [reply]
Could you elaborate why that would be difficult to manage ? Sounds pretty easy tbh

**Reply by BlackDex** [reply]
Maybe you can elaborate on your thoughts why it would be easy?
We might overlook something easy maybe.

**Reply by tamaskan** [reply]
I think you basically need to add the pgp-part to the send_email-function in mail.rs and add a email-route (like /admin) together with a link(bookmark-hint) to it in the welcome-email-template. In this template you would validate the user with username/password and offer a textbox for the key/checkboxes for options/save-button. When saving, the user would receive the encrypted email with a confirmation-link to activate these settings. Saving those settings could be done via a separate sqlite/sidecar-files if this solution should not touch the main database

**Reply by BlackDex** [reply]
That still doesn't provide a user with an interface where they can securely login and store there public keys to be used.
We either need to extract those keys from some gpg/pgp key provider, or, we need to use the one's provided.
Now, extracting the one's from a third party key provider seems nice, there is no guarantee it is the correct one.

Someone might want to use a totally different key then provided.
We can't create a nice login page for the user unless we try to use the exact same login code as Bitwarden does on a totally separate environment like we have now for `/admin`, when then need to use something like `/users` or whatever.

While it might be nice, it's not something i currently miss my self, or tend to work on in the near future.

**Reply by tamaskan** [reply]
That template would provide users a way to login and store the keys. Of course they have to either provide it or pull it from a provider. With that solution the only call to the api would be to validate the credentials.

**Reply by BlackDex** [reply]
Which template?

**Reply by tamaskan** [reply]
The page after the user logged in to the new eg /pgp-route. Sorry for the confusion

**Reply by BlackDex** [reply]
Well, that doesn't exists for one. Second, that should be build into the web-vault how you would suggest it, which is not something we want to do. We want to keep as close as possible to the upstream releases, and adding something like this would mean a lot of changes and adjustments.

**Reply by tamaskan** [reply]
It doesn't have to be build in, except the changes to the mail-function. I don't know rust, so i can't provide an example, sorry.

**Reply by BlackDex** [reply]
You mentioned a page after login/register, that is not something we controle, and thus need to change.
That whole part has nothing to do with Rust btw.

**Reply by tamaskan** [reply]
I guess the easiest way would be to fuse a custom sendmail with pgp like https://github.com/tamaskan/sendmail-pgp/tree/main and add a small node-website via proxy to handle the key-upload. That way there would be no need to touch the vaultwarden-code at all :-)

**Reply by BlackDex** [reply]
Well, nobody is stopping you from implementing something like that.

While i do like the feature, I do not think it will be part of Vaultwarden (or Bitwarden) in the near future.
If someone finds a nice way without changing the web-vault in a way that makes it unmaintainable when there are large changes done upstream, or even without changing the web-vault, but still have a secure way to verify the user is correct which uploads the key, then we are open for a well written PR.
